### ISSUE TITLE
intermittent SIGSEGV segfault when plotting during `training_step`

### ISSUE THREAD

**Issue by JackKelly** [PyTorchLightning/pytorch-lightning#11925]
## 🐛 Bug

If I create specific, simple `matplotlib` plots during `training_step` then I get a segfault after training for several batches.

Specifically, if I call `plot_data()` from within `training_step`, then I get a segfault:

```python
def plot_data():
    """Plot random data."""
    fig, axes = plt.subplots(ncols=4)  # See note 1 below.
    N_PERIODS = 16
    x = pd.date_range(START_DATE, periods=N_PERIODS, freq="30 min")  # See note 2 below.
    y = np.ones(N_PERIODS)
    for ax in axes:
        ax.plot(x, y)
    plt.close(fig)
```

Interestingly, I've found two ways to stop the segfaults. You can do one or the other (or both) of these actions to stop the segfaults:

1. Change `ncols` to one of `{2, 3, 5, 6, 7, 9, 10}`.  segfaults only seem to appear if `ncols` is one of `{4, 8, 16, 32}`
2. Convert `x` from a `pd.DatetimeIndex` to a numpy array of matplotlib dates by doing `x = matplotlib.dates.date2num(x)` after line 1.

If I run `gdb --args python script.py` I get this:

```
Epoch 0:   1%|█▏| 7/1024 [00:01<02:52,  5.90it/s, loss=0.281, v_num=34]
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x000055555568bb4b in _PyObject_GC_UNTRACK_impl (filename=0x55555587fef0 
"/home/conda/feedstock_root/build_artifacts/python-split_1643749964416/work/Modules/gcmodule.c", lineno=2236, 
op=0x7fffd05f5e00) at /home/conda/feedstock_root/build_artifacts/python-
split_1643749964416/work/Include/internal/pycore_object.h:76
76  /home/conda/feedstock_root/build_artifacts/python-split_1643749964416/work/Include/internal/pycore_object.h:
 No such file or directory.
```

The minimal code example below seems to always crash whilst training on epoch 7.

In my "real" code, the crash was intermittent.

If `ncols` is large (e.g. 32) then the crash changes randomly between:

* `SIGSEGV` segfault
* `AttributeError: 'weakref' object has no attribute 'grad_fn'`
* `AttributeError: 'builtin_function_or_method' object has no attribute 'grad_fn'`

The two `AttributeErrors` originate from the same bit of code: line 213 of `torch/optim/optimizer.py`: `if p.grad.grad_fn is not None:`.  My guess is that the `SIGSEGV` is the root problem, and the `AttributeErrors` are _symptoms_ of the underlying `SIGSEGV` error. I have (once) seen the code compain, at the same time, about a `SIGSEGV` and an `AttributeError`.

### To Reproduce

```python
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import pandas as pd
import numpy as np
import torch
from torch import nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import pytorch_lightning as pl


START_DATE = pd.Timestamp("2020-01-01")
N_EXAMPLES_PER_BATCH = 32
N_FEATURES = 1


class MyDataset(Dataset):
    def __init__(self, n_batches_per_epoch: int):
        self.n_batches_per_epoch = n_batches_per_epoch

    def __len__(self):
        return self.n_batches_per_epoch

    def __getitem__(self, idx: int) -> tuple[torch.Tensor, torch.Tensor]:
        x = torch.rand(N_EXAMPLES_PER_BATCH, N_FEATURES, 1)
        y = torch.rand(N_EXAMPLES_PER_BATCH, 1)
        return x, y


class LitNeuralNetwork(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.flatten = nn.Flatten()
        self.nn = nn.Linear(in_features=N_FEATURES, out_features=1)

    def forward(self, x):
        x = self.flatten(x)
        return self.nn(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.mse_loss(y_hat, y)
        plot_data()
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.001)


def plot_data():
    """Plot random data."""
    # ncols needs to be 4 or higher to trigger a segfault.
    fig, axes = plt.subplots(ncols=4)
    N_PERIODS = 16
    x = pd.date_range(START_DATE, periods=N_PERIODS, freq="30 min")
    # The segfaults go away if I do:
    # x = mdates.date2num(x)
    y = np.ones(N_PERIODS)
    for ax in axes:
        ax.plot(x, y)
    plt.close(fig)


dataloader = DataLoader(
    MyDataset(n_batches_per_epoch=1024),
    batch_size=None,
    num_workers=2,
)

model = LitNeuralNetwork()
trainer = pl.Trainer()
trainer.fit(model=model, train_dataloader=dataloader)
```

### Expected behavior

No segfault :slightly_smiling_face: 

### Environment

```
* CUDA:
	- GPU:
	- available:         False
	- version:           None
* Packages:
	- numpy:             1.22.2
	- pyTorch_debug:     False
	- pyTorch_version:   1.10.2
	- pytorch-lightning: 1.5.10
	- tqdm:              4.62.3
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.10
	- version:           #31-Ubuntu SMP Thu Jan 13 17:41:06 UTC 2022
```

Here's my `environment.yml` file:

```yaml
name: segfault
channels:
  - conda-forge
  - pytorch
dependencies:
  - matplotlib
  - pandas
  - pytorch
  - cpuonly 
  - pytorch-lightning
```

### Additional context

This may be related to issue #11067.

I have tried and failed to reproduce this problem in two simpler scripts:
* A script which _just_ uses matplotlib and pandas (no pytorch.  no pytorch-lightning)
* A script which _just_ uses matplotlib, pandas and pytorch (no pytorch-lightning).

These scripts are in a tiny GitHub repo for this bug report: https://github.com/JackKelly/pytorch-lighting-segfault

**Comment by justusschock** [comment]
@JackKelly Thanks for opening this issue and for the reproducibility examples. I'll have a look later today and report back if I find something :)

**Comment by justusschock** [comment]
Hi again,

I ran your code and for me, neither the pure pytorch, nor the lightning version segfaulted (or had any issue at all).

<img width="950" alt="image" src="https://user-images.githubusercontent.com/12886177/154113432-68d3e610-8c05-46e0-a383-939d03a2f2af.png">

The two attribute errors you mentioned seem to indicate that somewhere `torch.Tensor` are swapped out with proxies or builtin functions, for which I don't have any explanation.

I am very sorry, but with your exact environment, the code seems to run fine for me so I am a bit clueless on how to investigate this :/

**Comment by JackKelly** [comment]
Thanks loads for looking into this so quickly!

Hmm, that is strange!

Would it be possible to try increasing `ncols` to 32 (on the line `fig, axes = plt.subplots(ncols=4)`) to see if that forces a segfault?

I've seen segfaults with this code on my local workstation and on GCP VMs.  (All were running very similar Ubuntu 21.10 environments though)

Let me try again with a fresh VM on GCP...

**Comment by JackKelly** [comment]
Hmm, interesting, after installing a fresh VM, now I'm seeing the same as you! Here are the steps I took to create a fresh VM... Let me have a tinker to see if I can figure out what's going on!

1. [Create a new GCP VM instance](https://console.cloud.google.com/compute/instancesAdd)
2. Use machine config `e2-standard-4 (4 vCPU, 16 GB memory)`.  No GPU.
3. Change the boot disk to Ubuntu 21.10
4. Save and start the VM
5. SSH into the new VM
6. `sudo apt update && sudo apt -y upgrade`
7. [install miniconda](https://docs.conda.io/en/latest/miniconda.html)
8. log out and log back in again
9. `git clone https://github.com/JackKelly/pytorch-lighting-segfault.git`
10. `cd pytorch-lightning-segfault`
11. `conda env create -f environment.yml`
12. `conda activate segfault`
13. `python pytorch-lightning.py`

**Comment by JackKelly** [comment]
oops, sorry, I'm an idiot!

I had forgotten to comment out `x = mdates.date2num(x)`. I have now commented out `x = mdates.date2num(x)` and pushed to the repo.

Please pull from [the repo](https://github.com/JackKelly/pytorch-lighting-segfault) again and try running `python pytorch-lightning.py`.

It segfaults for me on the fresh VM (described in the comment above) and on my local workstation.

**Comment by justusschock** [comment]
Hi, Thanks for the update. Now, I can indeed reproduce this.

I changed your script to the following (which is functionally equivalent):

```python
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
import pandas as pd
import numpy as np
import torch
from torch import nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader
import pytorch_lightning as pl


START_DATE = pd.Timestamp("2020-01-01")
N_EXAMPLES_PER_BATCH = 32
N_FEATURES = 1


class MyDataset(Dataset):
    def __init__(self, n_batches_per_epoch: int):
        self.n_batches_per_epoch = n_batches_per_epoch

    def __len__(self):
        return self.n_batches_per_epoch

    def __getitem__(self, idx: int) -> tuple[torch.Tensor, torch.Tensor]:
        x = torch.rand(N_EXAMPLES_PER_BATCH, N_FEATURES, 1)
        y = torch.rand(N_EXAMPLES_PER_BATCH, 1)
        return x, y


class LitNeuralNetwork(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.flatten = nn.Flatten()
        self.nn = nn.Linear(in_features=N_FEATURES, out_features=1)

    def forward(self, x):
        x = self.flatten(x)
        return self.nn(x)

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.mse_loss(y_hat, y)
        from copy import deepcopy
        old_params = deepcopy(list(self.parameters()))
        self.plot_data()
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.001)


    def plot_data(self):
        """Plot random data."""
        # ncols which cause segfaults: 4, 8, 16, 32
        # ncols which don't appear to cause segfaults: 2, 3, 5, 6, 7, 9, 10
        from copy import deepcopy
        old_params = deepcopy(list(self.parameters()))
        fig, axes = plt.subplots(ncols=32)
        N_PERIODS = 16
        x = pd.date_range(START_DATE, periods=N_PERIODS, freq="30 min")
        # The segfaults go away if I do:
        # x = mdates.date2num(x)
        y = np.ones(N_PERIODS)
        for ax in axes:
            ax.plot(x, y)
        plt.close(fig)


dataloader = DataLoader(
    MyDataset(n_batches_per_epoch=1024),
    batch_size=None,
    num_workers=2,
)

model = LitNeuralNetwork()
trainer = pl.Trainer()
trainer.fit(model=model, train_dataloader=dataloader)
```

and when setting breakpoints with a condition of 
```python 
any(p.grad is not None and not isinstance(p.grad, torch.Tensor) for p in self.parameters())
```

I can see, that if it happens, it is after the `plt.subplots` command. Comparing the grad attribute before and after this command gives:

```python
>> for p in old_params:
>>     print(p.grad)
None
None
>> for p in self.parameters():
>>     print(p.grad)
<built-in method pop of dict object at 0x7f44eca02540>
<weakref at 0x7f44d5510f90; to 'CompositeGenericTransform' at 0x7f44ec9e6cd0>
```

I wasn't able to really debug this further, because it quickly dispatches to the matplotlib backend which is dynamically imported. You could try to simply switch the backend there, which I couldn't try (since I'm only having a remote machine without X-Server. 

I strongly think that this is a bug in matplotlib, since they also dynamically pass over all the locals when instantiating their `figure_manager`. 

Beyond that, I am sorry that I cannot help further.

**Comment by JackKelly** [comment]
Cool, thank you so much for looking into this, and for doing some debugging!

Should I report this bug to the `matplotlib` folks? (If I do that, please may I quote your debugging results?)

**Comment by justusschock** [comment]
Yes, I think that would be best as I don't see any reason why this could be caused by us :)

Sure, feel free to quote them :)

**Comment by JackKelly** [comment]
Cool, thank you. I'll close this issue for now. I've started a new issue in the matplotlib issue queue: https://github.com/matplotlib/matplotlib/issues/22478

Thanks again for your help!

**Comment by tacaswell** [comment]
https://github.com/matplotlib/matplotlib/issues/22478#issuecomment-1043069215

@justusschock Can you check the types of the objects in `self.paramaters()`?  I am assuming that somewhere in the pytorch/lightning side there are c-pointers to Python objects being held.  If that is true, I suspect there is a bug in your reference counting and somehow resurrected or null objects are being found.  Quoting my self from the mpl issue:

>This is weird enough that my guess is that under the hood pytorch/pytoch-lightning is keeping pointers to python objects without properly reference counting them, the objects are getting cleaned up, and then because Python uses an object pool that object is getting resurrected as a new object. I do not have a clear explanation of how the grad is getting miss-mapped to some other attribute (but my guess is that Paramaters is slotted so there is a index-based jump-table in there someplace). This would certainly explain the segfaults in the case when the object did not get resurrected and now instead of grabbing the wrong attribute it is looking at a null pointer.

**Comment by tacaswell** [comment]
I would also add I do not see _any_ mechanisim by which Matplotlib would even be aware of the objects in `self.parameters()`, let alone mutate them.

**Comment by justusschock** [comment]
Hi, I just checked the issue. 

thanks for the pointers @tacaswell .

We are indeed doing some pointer mechanics under the hood, but nothing of that is in any sort related to gradient tracking or anything similar (and it is all in Python). Those are all handled by PyTorch directly. So from my understanding it is definitely possible/likely that PyTorch does have some C-/C++-Style Pointers there. 

As I really do not have any idea on how to further debug this, I'd actually suggest to ask the `torch` guys directly. We are never touching the gradients explicitly except through pre-defined `torch` functions. I am really sorry, that I have to keep the hot-potato-game up and running, but I think this is your best chance to get the issue resolved. At least I am very clueless about what is going on here.

@tacaswell Yes I agree, that matplotlib wouldn't know about the parameters, but my initial guess was, that there might be a collision from one of the matplotlib backends (which might also include some C++ / C code) that might have a collision with the PyTorch C++ part. I am not that familiar with the matplotlib mechanics under the hood though so this is a real wild guess and I am sorry if it actually does not make sense at all.

**Comment by tacaswell** [comment]
I think looking at the `id(p)` and `type(p)` of the objects in `self.parameters()` would be useful information to take to `torch`.  If you run that debugging code mulitple times do you always get the same result (pop and a weakref to a transfrom)?  Is the dict that `pop` belongs to the `p` or has a bound method been stashed on an object?

I also suggest replacing the mpl calls with (a couple of ) calls to `gc.collect()` to see if that will force the issue.

While I can not rule out reference counting bugs on the mpl side, I would be very surprised if we were grabbing someone else's objects and adding things to them (or transmute it to a dict?).

**Comment by justusschock** [comment]
@tacaswell yes, they are always the same result (`pop` and `weakref`-wise). The only difference is, if this will result in an `AttributeError` or a `SIGSEGV` as @JackKelly pointed out.

**Comment by carmocca** [comment]
Huh, I'm experiencing this error now locally trying to run the PyCharm debugger for a test we have with a specific set of breakpoints enabled... 😕 

The test is:
https://github.com/PyTorchLightning/pytorch-lightning/blob/550d3a640d1bfeb731a16b2996d3160f0f7eb071/tests/trainer/test_trainer.py#L1300-L1318

And my environment:

```python
* CUDA:
        - GPU:
        - available:         False
        - version:           None
* Packages:
        - numpy:             1.21.0
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.1
        - pytorch-lightning: 1.6.0dev
        - tqdm:              4.62.3
* System:
        - OS:                Darwin
        - architecture:
                - 64bit
                - 
        - processor:         i386
        - python:            3.9.1
        - version:           Darwin Kernel Version 19.6.0: Tue Oct 12 18:34:05 PDT 2021; root:xnu-6153.141.43~1/RELEASE_X86_64
```

I don't think this is a problem with Lightning. But I can see we both have the same PyTorch and Python version.

**Comment by carmocca** [comment]
Yes. The report at https://github.com/PyTorchLightning/pytorch-lightning/issues/11067 used Python 3.6 but also PyTorch 1.10 so that seems to be the common element atm

**Comment by tacaswell** [comment]
It might also be worth manually turning off cyclic gc.  It will burn a bunch of memory, but if the issue is something holding a pointer to a gc'd object it would "fix" the problem.

**Comment by felipegb94** [comment]
@JackKelly I encountering a similar segfault when trying to log an image with tensorboard inside the training step (i.e., I use  `self.logger.experiment.add_image` inside the training step) .

One thing that solves the segfault is adding a `plt.pause(0.1)` _right after_ I finish plotting or adding the image:

```
import matplotlib.pyplot as plt
plt.pause(0.1)
```

Since I only log the images a few times during the epoch, adding a pause does not impact the performance of my training code.
